man/man1/make.1
'\" t
.\".ds ~ ~
.\" @(#)make.1 1.1 92/07/30 SMI; from UCB 4.3 BSD
.TH MAKE 1 "15 September 1989"
.SH NAME
make \- maintain, update, and regenerate related programs and files
.SH SYNOPSIS
.B make
[
.BI \-f " makefile"
] \&.\|.\|.
.RB [ " \-d " ]
.RB [ " \-dd " ]
.\".RB [ " \-ddd " ]
.RB [ " \-D " ]
.RB [ " \-DD " ]
.\".RB [ " \-DDD " ]
.if n .ti +.5i
.RB [ " \-e " ]
.RB [ " \-i " ]
.RB [ " \-k " ]
.\".RB [ " \-l " ]
.RB [ " \-n " ]
.\".RB [ " \-N " ]
.RB [ " \-p " ]
.RB [ " \-P " ]
.RB [ " \-q " ]
.RB [ " \-r " ]
.if t .ti +.5i
.\".RB [ " \-R " ]
.if n .ti +.5i
.RB [ " \-s " ]
.RB [ " \-S " ]
.RB [ " \-t " ]
.\".if t .ti +.5i
.\"[
.\".BI \-M " machines-file"
.\"]
[
.I target
\&.\|.\|.
]
.if n .ti +.5i
[
.IB macro = value
\&.\|.\|.
]
.IX  "make command"  ""  "\fLmake\fP \(em build programs"  "" PAGE START
.IX  "programming tools"  make  ""  "\fLmake\fP \(em build programs"  PAGE START
.IX  "build programs make"  ""  "build programs \(em \fLmake\fP"  ""  PAGE START
.IX  "maintain programs make"  ""  "maintain programs \(em \fLmake\fP"  ""  PAGE START
.IX  "update programs make"  ""  "update programs \(em \fLmake\fP"  ""  PAGE START
.IX  "regenerate programs make"  ""  "regenerate programs \(em \fLmake\fP"  ""  PAGE START
.SH DESCRIPTION
.LP
.B make
executes a list of shell commands associated with each
.IR target ,
typically to create or update a file of the same name.
.I makefile
contains entries that describe how to bring a target
up to date with respect to those on which it depends, which are called
.IR dependencies .
Since each dependency is a target, it may have dependencies
of its own.
Targets, dependencies, and sub-dependencies comprise a tree
structure that
.B make
traces when deciding whether or not to rebuild a
.IR target .
.LP
.B make
recursively checks each
.I target
against its dependencies, beginning with the first target entry in
.I makefile
if no
.I target
argument is supplied on the command line.
If, after processing all of its dependencies, a target file is
found either to be missing, or to be older than any of its dependencies,
.B make
rebuilds it.  Optionally with this version of
.BR make ,
a target can be treated as out-of-date when the commands used to
generate it have changed since the last time the target was built.
.LP
To build a given target,
.B make
executes the list of commands, called a
.IR rule .
This rule may be listed explicitly in the target's
makefile entry, or it may be supplied implicitly by
.BR make .
.LP
When no
.I makefile
is specified with a
.B \-f
option:
.RS
.TP 3
\(bu
If there is a file named
.B makefile
in the working directory,
.B make
uses that file.  If, however, there is an
.SM SCCS
history file
(\fB\s-1SCCS\s0/s.makefile\fP)
which is newer,
.B make
attempts to retrieve and use the most recent version.
.TP
\(bu
In the absence of the above file(s), if a file named
.B Makefile
is present in the working directory,
.B make
attempts to use it.  If there is an
.SM SCCS
history file
(\fB\s-1SCCS\s0/s.Makefile\fP)
that is newer,
.B make
attempts to retrieve and use the most recent version.
.RE
.LP
If no
.I target
is specified on the command line,
.B make
uses the first target defined in
.IR makefile .
.LP
If a
.I target
has no makefile entry, or if its entry has no rule,
.B make
attempts to derive a rule by each of the following methods, in
turn, until a suitable rule is found.  (Each method is
described under
.SM USAGE
below.)
.RS
.TP 3
\(bu
Pattern matching rules.
.TP
\(bu
Implicit rules, read in from a user-supplied makefile.
.TP
\(bu
Standard implicit rules (also known as
suffix rules), typically read in from the file
.BR <make/default.mk> .
.TP
\(bu
.SM SCCS
retrieval.
.B make
retrieves the most recent version from the
.SM SCCS
history file (if any).  See the description of the
.SB .SCCS_GET:
special-function target for details.
.TP
\(bu
The rule from the
.SB .DEFAULT:
target entry, if there is such an entry in the makefile.
.RE
.LP
If there is no makefile entry for a
.IR target ,
if no rule can be
derived for building it, and if no file by that name is present,
.B make
issues an error message and halts.
.SH OPTIONS
.TP
.BI \-f " makefile"
Use the description file
.IR makefile .
A
.RB ` \- '
as the
.I makefile
argument denotes the standard input.
The contents of
.IR makefile ,
when present, override the standard set of implicit rules and
predefined macros.  When more than one
.RB ` "\-f \fI\ makefile\fP" '
argument pair appears,
.B make
uses the concatenation of those files, in order of appearance.
.TP
.B \-d
Display the reasons why
.B make
chooses to rebuild a target;
.B make
displays any and all dependencies that are newer.
In addition,
.B make
displays options read in from the
.SB MAKEFLAGS
environment variable.
.TP
.B \-dd
Display the dependency check and processing in vast detail.
.TP
.B \-D
Display the text of the makefiles read in.
.TP
.B \-DD
Display the text of the makefiles,
.B default.mk
file, the state file, and all hidden-dependency reports.
.TP
.B \-e
Environment variables override assignments within makefiles.
.TP
.B \-i
Ignore error codes returned by commands.  Equivalent to the
special-function target
.RB ` \s-1.IGNORE:\s0 '.
.TP
.B \-k
When a nonzero error status is returned by a rule, or when
.B make
cannot find a rule,
abandon work on the current target, but continue with other dependency
branches that do not depend on it.
.\".TP
.\".B \-l
.\"Loop through targets interactively.
.\".B make
.\"reads the makefile, and requests a target to build.  When
.\"finished with a target,
.\".B make
.\"requests another, until you respond with
.\".BR quit .
.TP
.B \-n
No execution mode.
Print commands, but do not execute them.
Even lines beginning with an
.B @
are printed.  However, if a command line contains a reference to the
.B $(\s-1MAKE\s+1)
macro, that line is always executed (see the discussion of
.SB MAKEFLAGS
in
.BR "Reading Makefiles and the Environment" ).
.\".TP
.\".B \-N
.\"Override
.\".B \-n
.\"to force execution.
.TP
.B \-p
Print out the complete set of macro
definitions and target descriptions.
.\".TP
.\".B \-P
.\"Report dependencies for the target or targets, without rebuilding them.
.TP
.B \-P
Merely report dependencies, rather than building them.
.TP
.B \-q
Question mode.
.B make
returns a zero or nonzero status code
depending on whether or not the target file is up to date.
.TP
.B \-r
Do not read in the default makefile
.BR <make/default.mk> .
.\".TP
.\".B \-R
.\"Do not distribute targets nor build in parallel.  Normally, if the
.\".BI .make.machines
.\"file is found in your home directory, the machines defined in that file are
.\"used to rebuild targets that require it.
.TP
.B \-s
Silent mode.
Do not print command lines before executing them.
Equivalent to the special-function target
.RB ` \&.\s-1SILENT:\s0 '.
.TP
.B \-S
Undo the effect of the
.B \-k
option.
Stop processing when a non-zero exit status is returned
by a command.
.TP
.B \-t
Touch the target files (bringing them up
to date) rather than performing
their rules.
.ft I
This can be dangerous when files are
maintained by more than one person.
.ft R
When the
.SB \&.KEEP_STATE:
target appears in the makefile, this option updates the state
file just as if the rules had been performed.
.\".TP
.\".B \-w
.\"Report the working directory if a fatal error occurs.  This comes in
.\"handy for debugging recursive makefiles.
.\".TP
.\".B \-ww
.\"Report the working directory before running each command.
.\".TP
.\".B \-b
.\"This option has no effect, but is present for compatibility reasons.
.\".TP
.\".BI \-M " machines-file"
.\"Use the
.\".IR "machines-file"
.\"to define the set of machines to use for distributing target builds.
.\"By default,
.\".B make
.\"normally reads the file
.\".BR .make.machines
.\"in your home directory.
.TP
.IB macro = value
Macro definition.
This definition overrides any regular definition for the specified
macro within the makefile itself, or in the environment.  However,
this definition can still be overridden by conditional macro
assignments.
.SH USAGE
.LP
Refer to
.TX DMBG
and
.B make
in
.TX PUL
for tutorial information about
.B make.
.SS "Reading Makefiles and the Environment"
.LP
When
.B make
first starts, it reads the
.SB MAKEFLAGS
environment variable to obtain any the following
options specified present in its value:
.BR \-d ,
.BR \-D ,
.BR \-e ,
.BR \-i ,
.BR \-k ,
.BR \-l ,
.BR \-n ,
.BR \-p ,
.BR \-q ,
.BR \-r ,
.BR \-s ,
.BR \-S ,
or
.BR \-t .
(Within the
.SB MAKEFLAGS
value, the leading
.RB ` \(em '
character for the option string is omitted.)
.B make
then reads the command line for additional
options, which also take effect.
.LP
Next,
.B make
reads in a default makefile that typically contains predefined
macro definitions, target entries for implicit rules, and
additional rules, such as the rule for retrieving
.SM SCCS
files.
If present,
.B make
uses the file
.B default.mk
in the current directory; otherwise it reads the file
.BR <make/default.mk> ,
which contains the standard definitions and rules.
.br
.ne 4
Use the directive:
.IP
.B include <make/default.mk>
.LP
in your local
.B default.mk
file to include them.
.br
.ne 5
.LP
Next,
.B make
imports variables from the environment (unless the
.B \-e
option is in effect), and treats them as defined macros.  Because
.B make
uses the most recent definition it encounters, a macro definition
in the makefile normally overrides an
environment variable of the same name.
When
.B \-e
is in effect, however, environment variables are read in
.I after
all makefiles have been read.
In that case, the environment variables
take precedence over definitions in the makefile.
.LP
Next,
.B make
reads the state file,
.B .make.state
in the local directory if it exists, and then
any makefiles you specify with
.BR \-f ,
or one of
.B makefile
or
.B Makefile
as described above.
.LP
Next, (after reading the environment if
.B \-e
is in effect),
.B make
reads in any macro definitions supplied as command line
arguments.  These override macro definitions in the makefile
and the environment both,
but only for the
.B make
command itself.
.LP
.B make
exports environment variables, using the most recently defined value.
Macro definitions supplied on the command line are not normally
exported, unless the macro is also an environment variable.
.LP
.B make
does not export macros defined in the makefile.
If an environment variable is set, and a macro with the same name
is defined on the command line,
.B make
exports its value as defined on the command line. 
Unless
.B \-e
is in effect, macro definitions within
the makefile take precedence over
those imported from the environment.
.LP
The macros
.BR \s-1MAKEFLAGS\s0 ,
.SM
.BR MAKE ,
.SM
.BR KEEP_STATE ,
.SM
.BR SHELL ,
.SM
.BR HOST_ARCH ,
.SM
.BR TARGET_ARCH ,
.SM
.BR HOST_MACH ,
and
.SM
.BR TARGET_MACH
are special cases. 
See
.B Special-Purpose Macros
below, for details.
.SS "Makefile Target Entries"
.LP
A target entry has the following format:
.RS
.HP
.IR target .\|.\|.
.RB [ : \||\| :: ]
.RI [ dependency "] .\|.\|."
.RB [ ;
.IR command "] .\|.\|."
.nf
.RI [ command ]
\&.\|.\|.
.fi
.RE
.LP
The first line contains the name of a target, or a space-separated
list of target names, terminated with a colon
or double colon.  If a list of targets is given, this is
equivalent to having a separate entry of the same form
for each target.
The colon(s) may be followed by a
.IR dependency ,
or a dependency list.
.B make
checks this list before building the target.
The dependency list may be terminated with a semicolon
.RB ( ; ),
which in turn can be followed by a single Bourne shell command.
Subsequent lines in the target entry begin with a
.SM TAB\s0,
and contain Bourne shell commands.
These commands comprise the rule for building the target.
.LP
Shell commands may be continued across input lines by escaping the
.SM NEWLINE
with a backslash
.RB ( \e ).
The continuing line must also start with a
.SM TAB\s0.
.LP
To rebuild a target,
.B make
expands macros, strips off initial
.SM TAB\s0
characters and either executes the command directly (if it contains no
shell metacharacters), or passes each command line to a Bourne shell
for execution.
.LP
The first line that does not begin with a
.SM TAB
or
.B #
begins another target or macro definition.
.SS "Makefile Special Characters"
.SS \fIGlobal\fP
.LP
.TP
.B #
Start a comment.  The comment ends at the next
.SM NEWLINE\s0.
If the
.B #
follows the
.SM TAB
in a command line, that line is passed to the shell (which also treats
.B #
as the start of a comment).
.TP
.BI include " filename"
If the word
.B include
appears as the first seven letters of a line and is followed by a
.SM SPACE
or
.SM TAB\s0,
the string that follows
is taken as a filename to interpolate at that line.
.B include
files can be nested to a depth of no more than about 16.  If
.I filename
is a macro reference, it is expanded.
.br
.ne 5
.SS "\fITargets and Dependencies\fR"
.LP
.TP
.B :
Target list terminator.
Words following the colon are
added to the dependency list for the target or targets. 
If a target is named in more than one colon-terminated target entry,
the dependencies for all its entries are added to form that
target's complete dependency list.
.br
.ne 7
.TP
.B ::
Target terminator for alternate dependencies.
When used in place of a
.RB ` : '
the double-colon allows a target
to be checked and updated with respect
to alternate dependency lists.  When the target is out-of-date
with respect to dependencies listed in the first alternate, it is built
according to the rule for that entry. 
When out-of-date with respect to dependencies in another alternate,
it is built according the rule in that other entry.
Implicit rules do not apply to double-colon targets;
you must supply a rule for each entry.  If no dependencies are
specified, the rule is always performed.
.HP
.I target
.RB [ "+ "
.IR target .\|.\|.\|]
.B :
.br
Target group.
The rule in the target entry builds all the indicated targets as a
group.  It is normally performed only once per
.B make
run, but is checked for command dependencies every time a target
in the group is encountered in the dependency scan.
.TP
.B %
Pattern matching wild card metacharacter.  Like the
.RB ` * '
shell wild card,
.RB ` % '
matches any string of zero or more characters in a
target name or dependency, in the target portion of a
conditional macro definition, or within a pattern replacement macro
reference.  Note: only one
.RB ` % '
can appear in a target, dependency-name, or pattern-replacement macro
reference.
.TP
.BI \&./ pathname
.B make
ignores the leading
.RB ` \&./ '
characters from targets with names given as pathnames relative to
\*(lqdot,\*(rq the working directory.
.SS "\fIMacros\fR"
.LP
.TP
.B =
Macro definition.
The word to the left of this character is
the macro name; words to the right comprise its value.
Leading and trailing white space characters are stripped from
the value.
A word break following the
.B =
is implied.
.TP
.B $
Macro reference.
The following character, or the parenthesized or
bracketed string, is interpreted as a macro reference:
.B make
expands the reference (including the
.BR $ )
by replacing it with the macro's value.
.TP
.PD 0
.B ( )
.TP
.B { }
Macro-reference name delimiters. 
A parenthesized or bracketed word appended to a
.B $
is taken as the name of the macro being referred to.
Without the delimiters,
.B make
recognizes only the first character as the macro name.
.PD
.TP
.B $$
A reference to the dollar-sign macro, the value of which is the
character
.RB ` $ '.
Used to pass variable expressions beginning with
.B $
to the shell, to refer to environment variables which are expanded by
the shell, or to delay processing of dynamic macros within the
dependency list of a target, until that target is actually processed.
.TP
.B \e$
Escaped dollar-sign character.  Interpreted as a literal dollar sign
within a rule.
.TP
.B +=
When used in place of
.RB ` = ',
appends a string to a macro definition (must be surrounded by
white space, unlike
.RB ` = ').
.TP
.B :=
Conditional macro assignment.
When preceded by a list of
targets with explicit target entries, the
macro definition that follows takes effect when processing only those
targets, and their dependencies.
.TP
.B :sh =
Define the value of a macro to be the output of a command (see
.BR "Command Substitutions" ,
below).
.TP
.B :sh
In a macro reference, execute the command stored in the macro,
and replace the reference with the output of that command (see
.BR "Command Substitutions" ).
.br
.ne 3
.SS \fIRules\fR
.TP
.B \- 
.B make
ignores any nonzero error code returned by a command line for which the
first non-\s-1TAB\s0 character is a
.RB ` \- '.
This character is not passed to the shell
as part of the command line.
.B make
normally terminates when a command returns
nonzero status, unless the
.B \-i
or
.B \-k
options, or the
.SB \&.IGNORE:
special-function target is in effect.
.br
.ne 3
.TP
.B @
If the first non-\s-1TAB\s0 character is a
.BR @ ,
.B make
does not print the command line before executing it.
This character is not passed to the shell.
.br
.ne 3
.TP
.B ?
Escape command-dependency checking.
Command lines starting with this
character are not subject to command dependency checking.
.TP
.B !
Force command-dependency checking. 
Command-dependency checking
is applied to command lines for which it
would otherwise be suppressed.
This checking is normally suppressed for lines that contain
references to the
.RB ` ? '
dynamic macro (for example,
.RB ` $? ').
.\".TP
.\".BI = macro = value
.\"Delayed macro definition.
.\"Temporarily defines
.\".I macro
.\"to have the given
.\".I value
.\"for all targets that are subsequently processed.
.LP
When any combination of
.RB ` \- ',
.RB ` @ ',
.RB ` ? ',
or
.RB ` ! '
.\".RB ` = ',
appear as the first characters after the
.SM TAB\s0,
all that are present apply.  None are passed to the shell.
.br
.ne 5
.SS "Special-Function Targets"
.LP
When incorporated in a makefile, the following target names perform
special-functions:
.\".TP 12
.\".SB \&.AR_REPLACE
.\"This target can be used to specify a rule to preprocess
.\"member files before updating an
.\".BR ar (1V)
.\"library.
.TP
.SB \&.DEFAULT:
If it has an entry in the makefile,
the rule for this target is used to
process a target when there is no other entry for it, no rule for
building it, and no
.SM SCCS
history file from which to retrieve a current version.
.B make
ignores any dependencies for this target.
.br
.ne 3
.TP
.SB \&.DONE:
If defined in the makefile,
.B make
processes this target and its dependencies
after all other targets are built.  This target is also performed 
when
.B make
halts with an error, unless the
.SB \&.FAILED
target is defined.
.TP
.SB \&.FAILED:
This target, along with its dependencies, is performed instead of
.SB \&.DONE
when defined in the makefile and
.B make
halts with an error.
.TP
.SB \&.IGNORE:
Ignore errors.
When this target appears in the makefile,
.B make
ignores non-zero error codes returned from commands.
.TP
.SB \&.INIT:
If defined in the makefile, this target and its dependencies are
built before any other targets are processed.
.TP
.SB \&.KEEP_STATE:
If this target appears in the makefile,
.B make
updates the state file,
.BR .make.state ,
in the current directory.  This target
also activates command dependencies, and hidden dependency checks.
.TP
.SB \&.MAKE_VERSION:
A target-entry of the form:
.RS
.IP
.BI \&\s-1.MAKE_VERSION:\s0\0\0\s-1VERSION\-\s0 number
.RE
.IP
enables version checking.  If the version of
.B make
differs from the version indicated,
.B make
issues a warning message.
.TP
.B \&\s-1.NO_PARALLEL:\s+1
Currently, this target has no effect, it is, however, reserved for
future use.
.\"List of targets not to be built in parallel with any other target.  If
.\"no dependency list is given, all targets will default to non-parallel (see
.\".BR "Building Targets in Parallel" ).
.TP
.B \&\s-1.PARALLEL:\s+1
Currently of no effect, but reserved for future use.
.\"List of targets to be built in parallel.  This can be used to indicate
.\"that some nested invocations of
.\".B make
.\"should run in parallel, or that some targets should be built in parallel
.\"after a
.\".SB .NO_PARALLEL:
.\"target has caused all targets to default to non-parallel (see
.\".BR "Building Targets in Parallel" ).
.TP
.B \&\s-1.PRECIOUS:\s+1
List of files not to delete.
.B make
does not remove any of the files listed as dependencies for this
target when interrupted.
.B make
normally removes the current target when it receives an interrupt.
.br
.ne 3
.TP
.SB \&.SCCS_GET:
This target contains the rule for retrieving the current version
of an
.SM SCCS
file from its history file.  To suppress automatic retrieval,
add an entry for this target with an empty rule to your makefile.
.br
.ne 3
.TP
.SB \&.SILENT:
Run silently. 
When this target appears in the makefile,
.B make
does not echo commands before executing them.
.TP
.SB \&.SUFFIXES:
The suffixes list for selecting implicit rules (see
.BR "The Suffixes List" ).
.br
.ne 5
.TP
.SB \&.WAIT:
Currently of no effect, but reserved for future use.
.\"When this special target appears in a dependency list and the dependents
.\"are being built in parallel, all dependents prior to the
.\".SB \&.WAIT
.\"target are finished before continuing with the dependency list.  This is
.\"used when some dependents must be completely built before others are
.\"started (see
.\".BR "Building Targets in Parallel" ).
.br
.ne 8
.SS "\fIClearing Special Targets\fP"
In this version of
.BR make ,
you can clear the definition of the following special targets by
supplying entries for them with no dependencies and no rule:
.LP
.RS
.BR \s-1.DEFAULT\s0 ,\ 
.BR \s-1.SCCS_GET\s0 ,\ 
and
.SB .SUFFIXES
.ps
.ft B
.ft
.RE
.SS Command Dependencies
.LP
When the
.SB .KEEP_STATE:
target appears in the makefile,
.B make
checks the command for building a target against the state file,
.BR \&.make.state .
If the command has changed since the last
.B make
run,
.B make
rebuilds the target.
.SS Hidden Dependencies
.LP
When the
.SB .KEEP_STATE:
target appears in the makefile,
.B make
reads reports from
.BR cpp (1)
and other compilation processors for any \*(lqhidden\*(rq files, such as
.B #include
files.  If the target is out of date with
respect to any of these files,
.B make
rebuilds it.
.\".SS "Symbolic Link Support"
.\".LP
.\".B make
.\"recognizes a target entry of the form:
.\".IP
.\".IB target ": \&\s-1.SYM_LINK_TO\s0 " dependency-file
.\".LP
.\"as an entry to maintain
.\".I target
.\"as a symbolic link to
.\".IR dependency-file .
.\".SB \&.SYM_LINK_TO
.\"is a special dependency that is only meaningful when it appears
.\"first in the dependency list, followed by a single
.\".IR dependency-file .
.\".B make
.\"processes
.\".I dependency-file
.\"as a normal dependency.  If
.\".I target
.\"exists and is not a symbolic link to
.\".IR dependency-file ,
.\".B make
.\"issues an error message and stops.
.\"If there is a rule in the target entry,
.\".B make
.\"issues a warning and ignores that rule.
.SS Macros
.LP
Entries of the form
.IP
.IB macro = value
.LP
define macros.
.I macro
is the name of the macro, and
.IR value ,
which consists of all characters up to a
comment character or unescaped
.SM NEWLINE\s0,
is the value.
.B make
strips both leading and trailing white space in accepting the value.
.LP
Subsequent references to the macro, of the forms:
.BI $( name )
or
.BI ${ name }
are replaced by
.IR value .
The parentheses or brackets can be omitted in a reference to a
macro with a single-character name.
.LP
Macro references can contain references
to other macros, in which case
nested references are expanded first.
.\"On lines in which the first
.\"character is a
.\".BR ! ,
.\"references contained in the definition are evaluated just before an
.\"initial assignment (as the makefile is being read).
.SS "\fISuffix Replacement Macro References\fR"
.LP
Substitutions within macros can be made as follows:
.IP
.BI $( name : string1 = string2)
.LP
where
.I string1
is either a suffix, or a word to be replaced in the macro
definition, and
.I string2
is the replacement suffix or word.  Words in a macro value are
separated by
.SM SPACE\s0,
.SM TAB\s0,
and escaped
.SM NEWLINE
characters.
.br
.ne 12
.SS "\fIPattern Replacement Macro References\fR"
.LP
Pattern matching replacements can also be applied to macros, with a
reference of the form:
.sp .5
.RS
.BI $( name :
.IB op % os =
.IB np % ns )
.RE
.LP
where
.I op
is the existing (old) prefix and
.I os
is the existing (old) suffix,
.I np
and
.I ns
are the new prefix and new suffix, respectively, and the pattern
matched by
.BR %
(a string of zero or more characters),
is carried forward from the value
being replaced.  For example:
.sp .5
.RS
.nf
.ft B
\s-1PROGRAM\s0=fabricate
\s-1DEBUG\s0= $(\s-1PROGRAM\s0:%=tmp/%\-g)
.fi
.ft R
.RE
.LP
sets the value of
.SB DEBUG
to
.BR tmp/fabricate\-g .
.LP
Note: pattern replacement macro references cannot be used
in the dependency list of a pattern matching rule; the
.B %
characters are not evaluated independently.  Also, any number of
.B %
metacharacters can appear after the equal-sign.
.br
.ne 5
.SS "\fIAppending to a Macro\fR"
.LP
Words can be appended to macro values as follows:
.IP
.IB macro " += " "word .\|.\|."
.br
.ne 5
.SS "Special-Purpose Macros"
.LP
When the
.SB MAKEFLAGS
variable is present in the environment,
.B make
takes options from it,
in combination with options entered on the command line.
.B make
retains this combined value as the
.SB MAKEFLAGS
macro, and exports it automatically to
each command or shell it invokes.
.LP
Note: flags passed by way of
.SB MAKEFLAGS
are only displayed when the
.BR \-d ,
or
.B \-dd
options are in effect.
.LP
The
.SB MAKE
macro is another special case.  It has the value
.B make
by default, and temporarily overrides the
.B \-n
option for any line in which it is referred to.
This allows nested invocations of
.B make
written as:
.IP
.BR $(\s-1MAKE\s0) " .\|.\|."
.LP
to run recursively, with the
.B \-n
flag in effect for all commands but
.B make.
This lets you use
.RB ` "make \-n" '
to test an entire hierarchy of makefiles.
.LP
For compatibility with the 4.2
.SM BSD
.BR make ,
the
.SB MFLAGS
macro is set from the
.SB MAKEFLAGS
variable by prepending a
.RB ` \- '.
.SB MFLAGS
is not exported automatically.
.LP
The
.SB SHELL
macro, when set to a single-word value such as
.BR /usr/bin/csh ,
indicates the name of an alternate shell to use.  The default is
.BR /bin/sh .
Note:
.B make
executes commands that contain no shell metacharacters itself.
Built-in commands, such as
.BR dirs
in the C shell, are not recognized unless the command line includes a
metacharacter (for instance, a semicolon).  This macro is neither
imported from, nor exported to the environment, regardless of
.BR \-e .
To be sure it is set properly,
you must define this macro within every
makefile that requires it.
.LP
The
.SB KEEP_STATE
environment variable has the same effect as the
.SB .KEEP_STATE:
special-function target.  It enables command dependencies, hidden
dependencies and writing of the state file.
.br
.ne 10
.LP

usr.etc/c2convert/make.c
/*
 *
 */

#ifndef lint
static  char sccsid[] = "@(#)make.c 1.1 92/07/30 Copyr 1987 Sun Micro"; /* c2 secure */
#endif

#include <stdio.h>

char *add_to_list();
char *list_reset();
char *list_next();
char *index();
char hostname[];

char *
make_fstabs(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/fstab", base);
	else
		sprintf(path, "/etc/fstab");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcat(path, "/etc/fstab");
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_audit_controls(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/security/audit/audit_control", base);
	else
		sprintf(path, "/etc/security/audit/audit_control");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcat(path, "/etc/security/audit/audit_control");
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_local_fs(devices)
	char *devices;
{
	char path[256];
	char *result = NULL;
	char *cp;
	int suffix = 0;

	for (cp = list_reset(devices); cp; cp = list_next(devices)) {
		if (suffix == 0)
			sprintf(path, "/dev/%s /etc/security/audit/%s",
			    cp, hostname);
		else
			sprintf(path, "/dev/%s /etc/security/audit/%s.%d",
			    cp, hostname, suffix);
		suffix++;
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_client_local_fs(local_fs)
	char *local_fs;
{
	char path[256];
	char *result = NULL;
	char *cp;
	char *bp;

	for (cp = list_reset(local_fs); cp; cp = list_next(local_fs)) {
		bp = index(cp, ' ') + 1;
		sprintf(path, "%s:%s %s", hostname, bp, bp);
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_remote_fs(filesystems)
	char *filesystems;
{
	char part[256];
	char path[256];
	char *result = NULL;
	char *cp;
	char *colon_p;

	for (cp = list_reset(filesystems); cp; cp = list_next(filesystems)) {

		if ((colon_p = index(cp, ':')) == NULL)
			sprintf(part, "%s:%s", hostname, cp);
		else
			sprintf(part, "%s", cp);
		free(cp);
		colon_p = index(part, ':');
		*colon_p++ = '\0'; 

		if (*colon_p == '/')
			sprintf(path, "%s:%s", part, colon_p);
		else
			sprintf(path, "%s:/etc/security/audit/%s", part,
			    colon_p);

		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_passwds(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char ppath[256];
	char sysstg[128];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/passwd", base);
	else
		sprintf(path, "/etc/passwd");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcpy(ppath, path) ;
		strcat(path, "/etc/passwd");
		free(cp);
		sprintf(sysstg,"grep '##' %s >/dev/null",path);
		if (system(sysstg) == 0) {
			printf("#\n# ## in password file of %s indicates C2 conversion previously run.\n#\n",path);
			printf("echo '## in password file of %s indicates C2 conversion previously run.'\n",path);
			exit(1);
			}
		sprintf(sysstg,"grep '^audit:' %s >/dev/null",path);

		if (system(sysstg) != 0) { /* somehow, no "audit"    */
					/* account in /etc/passwd */
					/* -- add one in	  */
			printf("#\n# must add 'audit' account to password file %s.\n#\n",path);
			printf("echo 'audit:*:9:9::/etc/security/audit:/bin/csh' >/tmp/C2.$$passwd\n");
			printf("cat %s >>/tmp/C2.$$passwd\n",path);
			printf("mv %s %s.bak\n", path, path);
			printf("mv /tmp/C2.$$passwd %s\n", path);
		}

        	add_pseudo_user(ppath,"AUyppasswdd", 28, 10);
		add_pseudo_user(ppath,"AUpwdauthd", 27, 10);

		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_passwd_adjuncts(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/security/passwd.adjunct", base);
	else
		sprintf(path, "/etc/security/passwd.adjunct");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcat(path, "/etc/security/passwd.adjunct");
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_groups(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/group", base);
	else
		sprintf(path, "/etc/group");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcat(path, "/etc/group");
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_group_adjuncts(base, roots, clients)
	char *base;
	char *roots;
	char *clients;
{
	char path[256];
	char *result;
	char *cp;

	if (base)
		sprintf(path, "%s/etc/security/group.adjunct", base);
	else
		sprintf(path, "/etc/security/group.adjunct");

	result = add_to_list(NULL, path);
	for (cp = list_reset(clients); cp; cp = list_next(clients)) {
		path[0] = '\0';
		if (base)
			strcat(path, base);
		if (roots)
			strcat(path, roots);
		strcat(path, "/");
		strcat(path, cp);
		strcat(path, "/etc/security/group.adjunct");
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_audit_warns(base, execs)
	char *base;
	char *execs;
{
	char path[256];
	char *result = NULL;
	char *cp;

	for (cp = list_reset(execs); cp; cp = list_next(execs)) {
		if (base)
			sprintf(path, "%s%s/etc/audit_warn", base, cp);
		else
			sprintf(path, "%s/etc/audit_warn", cp);
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}

char *
make_dir_list(directories)
	char *directories;
{
	char path[256];
	char *result = NULL;
	char *cp;

	for (cp = list_reset(directories); cp; cp = list_next(directories)) {
		sprintf(path, "%s", cp);
		free(cp);
		result = add_to_list(result, path);
	}
	return (result);
}
